### PR TITLE
Fix #304238: Prevent crash when using 'Image capture'

### PR DESCRIPTION
**Pull request by Jojo-Schmitz**
Solves https://musescore.org/en/node/304238

#5980 might be the better fix

### CODE REVIEW

**Comment by dmitrio95** [comment]
As for me, this one is better as it makes it more clear which coordinate systems all these offsets belong to, only making an exception for those elements which do not belong to any page. This is why I initially preferred to make a `findAncestor(ElementType)` function to use here instead of something like `findTopAncestor()`.

Also I have a feeling I have fixed this issue before, perhaps I somehow forgot to push the fix to my PR's branch.

File: mscore/editelement.cpp [hunk 1]
```diff
@@ -60,8 +60,8 @@ void ScoreView::updateGrips()
 
             // updateGrips returns grips in page coordinates,
             // transform to view coordinates:
-
-            const QPointF pageOffset(editData.element->findAncestor(ElementType::PAGE)->pos());
+            Element* e= editData.element->findAncestor(ElementType::PAGE);
+            const QPointF pageOffset(e ? e->pos() : QPointF());
```
[the remark this answers is not in the document]
Comment (Jojo-Schmitz) [reply mscore/editelement.cpp:64]: Yes, I could resize the lasso

[the remark this answers is not in the document]
Comment (Jojo-Schmitz) [reply mscore/editelement.cpp:64]: Ah, only once though :-(

[the remark this answers is not in the document]
Comment (Jojo-Schmitz) [reply mscore/editelement.cpp:64]: Only once **correctly** that is

**Comment by Jojo-Schmitz** [comment]
> In fact, this PR is already failing now because it's not correctly using the coordinate system of the element itself in the case where it's floating alone; as a result, the grips never appear.

The grips do appear, but only the first time at the right spot :-(

File: mscore/dragelement.cpp [hunk 2]
```diff
@@ -98,7 +98,8 @@ void ScoreView::doDragElement(QMouseEvent* ev)
 
       for (Element* e : sel.elements()) {
             QVector<QLineF> elAnchorLines = e->dragAnchorLines();
-            const QPointF pageOffset(e->findAncestor(ElementType::PAGE)->pos());
+            Element* el = e->findAncestor(ElementType::PAGE);
+            const QPointF pageOffset(el ? el->pos() : QPointF());
```
Comment (Spire42) [inline mscore/dragelement.cpp:102]: Change to:

`const QPointF pageOffset((page ? page : e)->pos());`

File: mscore/editelement.cpp [hunk 3]
```diff
@@ -60,8 +60,8 @@ void ScoreView::updateGrips()
 
             // updateGrips returns grips in page coordinates,
             // transform to view coordinates:
-
-            const QPointF pageOffset(editData.element->findAncestor(ElementType::PAGE)->pos());
+            Element* el = editData.element->findAncestor(ElementType::PAGE);
+            const QPointF pageOffset(el ? el->pos() : QPointF());
```
Comment (Spire42) [inline mscore/editelement.cpp:64]: Change to:

`const QPointF pageOffset((page ? page : editData.element)->pos());`

File: mscore/dragelement.cpp [hunk 4]
```diff
@@ -98,7 +98,8 @@ void ScoreView::doDragElement(QMouseEvent* ev)
 
       for (Element* e : sel.elements()) {
             QVector<QLineF> elAnchorLines = e->dragAnchorLines();
-            const QPointF pageOffset(e->findAncestor(ElementType::PAGE)->pos());
+            Element* el = e->findAncestor(ElementType::PAGE);
```
Comment (Spire42) [inline mscore/dragelement.cpp:101]: I suggest renaming `el` to `page` and making it `const`:

`Element* const page = e->findAncestor(ElementType::PAGE);`

File: mscore/editelement.cpp [hunk 5]
```diff
@@ -60,8 +60,8 @@ void ScoreView::updateGrips()
 
             // updateGrips returns grips in page coordinates,
             // transform to view coordinates:
-
-            const QPointF pageOffset(editData.element->findAncestor(ElementType::PAGE)->pos());
+            Element* el = editData.element->findAncestor(ElementType::PAGE);
```
Comment (Spire42) [inline mscore/editelement.cpp:63]: I suggest renaming `el` to `page` and making it `const`:

`Element* const page = editData.element->findAncestor(ElementType::PAGE);`

File: mscore/dragelement.cpp (hunk 2 above)
Comment (Jojo-Schmitz) [reply mscore/dragelement.cpp:102]: nope, doesn't even compile. But `const QPointF pageOffset((page() ? page() : e)->pos());` does ;-)

File: mscore/editelement.cpp (hunk 3 above)
Comment (Jojo-Schmitz) [reply mscore/editelement.cpp:64]: same here

**Comment by Jojo-Schmitz** [comment]
I still get it to crash though, in Debug mode, when draginng the selection around:
Fatal: ASSERT: "!isLocked()" in file ...\libmscore\select.cpp, line 327

**Comment by Spire42** [comment]
> I still get it to crash though, in Debug mode (QtCreator/MinGW), when dragging the selection around (a 2nd time):
> Fatal: ASSERT: "!isLocked()" in file ...\libmscore\select.cpp, line 327

Happens in my PR too. The cause is the code erroneously trying to select the capture-region element during the drag operation. I think this is a separate issue that possibly existed even before all of these recent changes.

**Comment by Jojo-Schmitz** [comment]
> Happens in my PR too. The cause is the code erroneously trying to select the capture-region element during the drag operation. I think this is a separate issue that possibly existed even before all of these recent changes.

Most probably, yes.

**Comment by igorkorsukov** [comment]
> I still get it to crash though, in Debug mode (QtCreator/MinGW), when dragging the selection around (a 2nd time):
> Fatal: ASSERT: "!isLocked()" in file ...\libmscore\select.cpp, line 327
> Doesn't happen with VS2019/MSVC and a RelWithDebInfo build though
> Maybe @igorkorsukov has an idea on this?

I don’t know the reasons for the fall, I need to investigate, but the assert is correct and it’s good that it worked, so they found an error.

Drag works like this:
1.start drag
```
for (Element* e : _score->selection().elements())
            e->startDrag(editData);
```
2. draging

3. end drag
```
for (Element* e : _score->selection().elements()) {
            e->endDrag(editData);
```

i.e. between the start and end of the drag the selection should not change!
To control the selection change, I added a lock function  

And now like this:
1.start drag
```

for (Element* e : _score->selection().elements())
            e->startDrag(editData);

_score->selection().lock("drag");
```
2. draging

3. end drag
```
for (Element* e : _score->selection().elements()) {
            e->endDrag(editData);
_score->selection().unlock("drag");
```

**Comment by Jojo-Schmitz** [comment]
I didn't mean to imply that the assert is wrong, just that we apparently found a way to trigger it ;-)

**Comment by igorkorsukov** [comment]
Users had crashes due to the fact that during the drag, the user changed the selection using the keyboard.  
Therefore I added
```
            {{"next-chord",
              "prev-chord",
              "next-track",
              "prev-track",
              "next-measure",
              "prev-measure"}, [](ScoreView* cv, const QByteArray& cmd) {

                  if (cv->score()->selection().isLocked()) {
                        LOGW() << "unable exec cmd: " << cmd << ", selection locked, reason: " << cv->score()->selection().lockReason();
                        return;
                        }
```

Probably something similar happens in this case.

**Comment by Jojo-Schmitz** [comment]
There isn't really a selection, so nothing to unselect, at least none that I can see?
Anyway, it doesn't crash in a Release or RelWithDebInfo build, so I guess we can get away with this PR as it is now?

File: mscore/dragelement.cpp (hunk 4 above)
Comment (Jojo-Schmitz) [reply mscore/dragelement.cpp]: much better name indeed

File: mscore/editelement.cpp (hunk 5 above)
Comment (Jojo-Schmitz) [reply mscore/editelement.cpp]: dito

File: mscore/dragelement.cpp (hunk 2 above)
Comment (Jojo-Schmitz) [reply mscore/dragelement.cpp]: And doing it this way indeed seems way better, thanks!

File: mscore/editelement.cpp (hunk 3 above)
Comment (Jojo-Schmitz) [reply mscore/editelement.cpp]: (same here)

**Comment by Spire42** [comment]
> Anyway, it doesn't crash in a Release or RelWithDebInfo build, so I guess we can get away with this PR as it is now?

Yes, but again, PR #5980 is a more general and safer solution. This one is just a band-aid. It's fragile because it has hard-coded special cases, and that fragility was demonstrated by the fact that one of those special cases was broken in the original commit.

**Comment by Jojo-Schmitz** [comment]
I don't mind which of those 2, both suffer from the same remaining crash under the very same circumstances

**Comment by dmitrio95** [comment]
> I get what you're saying, but this one is more fragile as it's hard-coded for two specific cases rather than presenting a general solution. What if in the future we have a floating element that has a parent and is using its parent's coordinate system?

Maybe I really rushed things too much with this change but in the future it would be better to come to something more consistent. The idea is, the offsets calculated here are used exclusively for translating coordinates returned by `dragAnchorLines()` and `gripsPositions()` to canvas coordinate system. These functions are supposed to return their results in page coordinates system, and elements which implement them are not supposed to return anything else there. Unfortunately I missed the fact this doesn't work for `Lasso` (although page coordinates were used for grips and anchor lines for a long time before these changes). Maybe a better solution for the future would be to make `gripsPositions()` and `dragAnchorLines()` return their results in canvas coordinates, or change coordinate systems handling in some other way. But until this is not changed to some consistent way probably both #5980 and this PR are acceptable, depending on how we would like to treat `gripsPositions()` and `dragAnchorLines()` for elements that do not belong to any page.

**Pull request merged**